File: Python/Questions/PairSum.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:

        # O(1) space solution O(n) time

        fast, slow, prev = head, head, None
        while fast and fast.next:
            fast = fast.next.next
            tmp = slow
            slow = slow.next
            tmp.next = prev
            prev = tmp

        max_count = 0

        while slow:
            curr_count = prev.val + slow.val
            max_count = max(max_count, curr_count)
            slow, prev = slow.next, prev.next
        
        return max_count
            





        # determine the middle point.

        # add every one of the points into a string

        """ The following approach is O(n) time and space """
        # slow, fast = head, head
        # while fast and fast.next:
        #     fast = fast.next.next
        #     slow = slow.next
        
        # arr = []
        # while slow:
        #     arr.append(slow.val)
        #     slow = slow.next
        
        # max_sum = 0
        # for n in reversed(arr):
        #     curr_sum = head.val + n
        #     max_sum = max(max_sum, curr_sum)
        #     head = head.next
        
        # return max_sum

File: c++/questions/pairSum.cpp
#include <vector>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    int pairSum(ListNode *head)
    {

        ListNode *fast = head, *slow = head, *prev = nullptr;
        while (fast && fast->next)
        {
            fast = fast->next->next;
            ListNode *tmp = slow;
            slow = slow->next;
            tmp->next = prev;
            prev = tmp;
        }

        int max_sum = 0; // unsigned ints
        while (slow)
        {
            int curr_sum = slow->val + prev->val;
            max_sum = max(max_sum, curr_sum);
            slow = slow->next;
            prev = prev->next;
        }

        return max_sum;
    }

    /* The following is the O(n) space and time solution */
    // add all of the values
    //     ListNode * slow = head, *fast = head;

    //     while (fast && fast->next) {
    //         fast = fast->next->next;
    //         slow = slow->next;
    //     }

    //     vector<int> arr;

    //     while (slow) {
    //         arr.push_back(slow->val);
    //         slow = slow->next;
    //     }

    //     int max_sum = 0;
    //     for (int i = arr.size() - 1; i >= 0; i--){
    //         int curr_sum = arr[i] + head->val;
    //         max_sum = max(max_sum, curr_sum);
    //         head = head->next;
    //     }

    //     return max_sum;
    // }
};